hie-plugin-api/Haskell/Ide/Engine/Context.hs

```diff
@@ -13,27 +15,86 @@ import Haskell.Ide.Engine.PluginUtils
 -- smarter code completion
 data Context = TypeContext
              | ValueContext
+             | ModuleContext String
+             | ImportContext String
+             | ImportListContext String
+             | ImportHidingContext String
+             | ExportContext
+             | InstanceContext
+             | ClassContext
+             | DerivingContext
   deriving (Show, Eq)
 
 -- | Generates a map of where the context is a type and where the context is a value
 -- i.e. where are the value decls and the type decls
 getContext :: Position -> ParsedModule -> Maybe Context
-getContext pos pm = everything join (Nothing `mkQ` go `extQ` goInline) decl
+getContext pos pm
+  | Just (L (RealSrcSpan r) modName) <- moduleHeader
+  , pos `isInsideRange` r
+  = Just (ModuleContext (moduleNameString modName))
+
+  | Just (L (RealSrcSpan r) _) <- exportList
+  , pos `isInsideRange` r
+  = Just ExportContext
+
+  | Just ctx <- everything (<|>) (Nothing `mkQ` go `extQ` goInline) decl
+  = Just ctx
+
+  | Just ctx <- asum $ map importGo imports
```

Comment:
I restrained myself to use `everything` here, since we do not wish to traverse every info in import lists, this seemed to be more efficient to me, but I do not understand `Data` enough, if it actually does what I would expect it to.

Reply:
Now uses `something` is the same as `everything (<|>)` afaik.